src/subdomains/analytics/application/services/staking-history.service.ts

```diff
@@ -45,13 +55,25 @@ export class StakingHistoryService {
     const withdrawals = await this.getWithdrawalsByUserOrAddress(userId, query.depositAddress, query.from, query.to);
     const rewards = await this.getRewardsByUserOrAddress(userId, query.depositAddress, query.from, query.to);
 
+    const uniqueTargetAddresses = withdrawals
+      .map((w) => w.staking.withdrawalAddress)
+      .concat(rewards.filter((r) => !r.isReinvest).map((r) => r.targetAddress))
+      .filter((a1, i, self) => i === self.findIndex((a2) => a1.isEqual(a2)));
+
+    const wallets = await this.walletService.getWalletsByAddresses(uniqueTargetAddresses);
+
+    const providerMap: WalletProviderAddressPair[] = wallets.map((w) => ({
+      targetAddress: w.address.address,
+      walletProvider: w.walletProvider.name,
+    }));
```

Comment:
I would use a real map instead of an array (makes it easier to access):
```ts
const exchangeMap = wallets.reduce(
      (map, w) => map.set(w.address.address, w.walletProvider.name),
      new Map<string, string>(),
    );
```